src/System.Linq.Expressions/src/System/Linq/Expressions/BlockExpression.cs

```diff
@@ -88,9 +88,37 @@ public override Type Type
         /// <returns>This expression if no children changed, or an expression with the updated children.</returns>
         public BlockExpression Update(IEnumerable<ParameterExpression> variables, IEnumerable<Expression> expressions)
         {
-            if (variables == Variables && expressions == Expressions)
+            bool variablesMatch;
+            if (variables == Variables)
+                variablesMatch = true;
+            else if (Variables == null || Variables.Count == 0)
+                variablesMatch = variables == null || !variables.Any();
+            else if (variables == null)
+                variablesMatch = false;
+            else
             {
-                return this;
+                // Just in case variables cannot be streams and cannot be repeated.
+                IList<ParameterExpression> varList = variables as IList<ParameterExpression> ?? variables.ToReadOnly();
+                variablesMatch = Variables.Count == varList.Count && Variables.SequenceEqual(varList);
+            }
+
+            if(variablesMatch)
+            {
+                bool expressionsMatch;
+                if (expressions == Expressions)
+                    expressionsMatch = true;
+                else if (expressions == null)
```

Comment:
You don't have the equivalent of 
```
else if (Variables == null || Variables.Count == 0)  
    variablesMatch = variables == null || !variables.Any(); 
```
for block expressions.  Is how the classes are used making that check unnecessary (always one block)?

I find myself twitching, wanting to consolidate the mostly identical checks of variables and expressions here (mostly to remove the nesting, because what I'm working though currently has too much of it).